### PR TITLE
fix: Add more precise types to Cypress.Commands

### LINKED ISSUE #18879: Typings for Command.add are missing types for the subject
### Current behavior

A simple custom command that relies on prevSubject is not correctly typed.

### Desired behavior

The "previous subject" argument should be correctly added to the command type and error if not passed.

### Test code to reproduce

```ts
declare global {
  namespace Cypress {
    interface Chainable {
      confirmClick(): void;
    }
  }
}

Cypress.Commands.add('confirmClick', { prevSubject: true }, (confirmButton) => {
  confirmButton.click();
  cy.wrap(confirmButton)
    .should('have.attr', 'data-state', 'confirming')
    .click({ force: true });
});

```

### Cypress Version

9.0.0

### Other

_No response_

### LINKED ISSUE #18940: `Cypress.Commands.add` type doesn't take into account `prevSubject` to define the type of the custom command
### Current behavior

```ts
declare global {
  namespace Cypress {
    interface Chainable {
      /**
       * Custom command to select DOM element by data-cy attribute.
       * @example cy.dataCy('greeting')
       */
      dataCy<E extends Node = HTMLElement>(value: string): Chainable<JQuery<E>>;
    }
  }
}

Cypress.Commands.add(
    'dataCy',
    { prevSubject: 'optional' },
    (subject, value) => {
      return cy.get(`[data-cy=${value}]`, {
        withinSubject: subject,
      });
    },
  );
```

`subject` and `value` are inferred as `any` as `prevSubject` isn't taken into account and TS expect a single param
If `value` is removed, the first param becomes `string` which is wrong
The same problem applies to child commands

### Desired behavior

Types should take into consideration `prevSubject` and use it into a conditional to provide correct typings
There may be other edge cases with `element` | `document` | `window`

### Test code to reproduce

https://github.com/cypress-io/cypress/pull/17496/files#diff-facdf706a5238380b5e97c040bc1a5f5cb89e3116f878754dd7fb4e693787715R77

This test is wrong, it should be

```ts
Cypress.Commands.add('newCommand', { prevSubject: true }, (subject, arg) => {
    // $ExpectType string
    arg
    return
  })
```

### Cypress Version

9.0.0

### Other

_No response_

### LINKED ISSUE #19095: `Cypress.Commands.overwrite` typescript problem appearing in 9.1
### Current behavior

Cypress 9 added type definitions to `Cypress.Commands.overwrite`, and I can't seem to make sense of them. 

This is what I have, in `cypress/support/commands.ts`: 

```typescript
if (Cypress.env('demoMode') === 'on') {
  // adapted from https://github.com/cypress-io/cypress/issues/249#issuecomment-670028947
  for (const commandName of ['visit', 'click', 'trigger', 'type', 'clear', 'reload'] as const) {
    // we add 2s delays for a few commands so that stakeholders can see what's going on
    const commandWithDelay = (command: (...args: unknown[]) => unknown, ...args: unknown[]) =>
      new Promise((resolve) => {
        setTimeout(() => resolve(command(...args)), 2000);
      });

    Cypress.Commands.overwrite(commandName, commandWithDelay);
  }
}
```

This was working fine until the new Cypress version. 

Let's tackle something simpler, taken out straight [from cypress docs](https://docs.cypress.io/api/cypress-api/custom-commands#Overwrite-type-command):

```typescript
Cypress.Commands.overwrite('type', (originalFn, element, text, options) => {
  if (options && options.sensitive) {
    // turn off original log
    options.log = false
    // create our own log with masked message
    Cypress.log({
      $el: element,
      name: 'type',
      message: '*'.repeat(text.length),
    })
  }

  return originalFn(element, text, options)
})
```

I can't get the types right, even for that seemingly simple example 😞 

### Desired behavior

_No response_

### Test code to reproduce

```typescript
Cypress.Commands.overwrite('type', (originalFn, element, text, options) => {
  if (options && options.sensitive) {
    // turn off original log
    options.log = false
    // create our own log with masked message
    Cypress.log({
      $el: element,
      name: 'type',
      message: '*'.repeat(text.length),
    })
  }

  return originalFn(element, text, options)
})
```

### Cypress Version

9.1

### Other

_No response_

### PR DESCRIPTION
**Pull request by xumepadismal**
- Closes #18879
- Closes #18940
- Closes #19095

### User facing changelog

* Custom command implementations typings take into account `prevSubject` variants
* Custom command implementations now allows to NOT return a value
* Custom command overwrites typings take into account `originalFn` function

### PR Tasks

- [x] Have tests been added/updated?
- [x] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [x] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [x] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
- N/A Have new configuration options been added to the [`cypress.schema.json`](https://github.com/cypress-io/cypress/blob/develop/cli/schema/cypress.schema.json)?

### CODE REVIEW

**Comment by emilyrohrbough** [comment]
@xumepadismal FYI - https://github.com/cypress-io/cypress/pull/18967 issued by @remcohaszing has similar types changes to Cypress.Commands as well. It looks like it takes prevSubject into account. Where you aware?

@remcohaszing Did you by chance come across this PR?

I'm curious what you both think of each other's proposed changes.

**Comment by xumepadismal** [comment]
@emilyrohrbough oh I missed these PRs indeed...

Here's my thoughts about the difference.

### Regarding the preSubject

PR #18880 seems to propose more future-proof approach to `prevSubject` typings because it supports other hypothetical options in `CommandOptions` besides `prevSubject`.

But on the other hand it doesn't support array of prevSubjects which IS supportred by this PR. You can refer to `cypress-tests.ts` in Changes of this PR to see the examples of how types are inferred.

### Regarding the void

In #18967 return type of `fn` is just `void` which is not accurate. We should allow `void` but not require it.
In my PR it is `ReturnType<ChainableMethods[T]> | void` which is better correspond the docs.

### ChainableMethods vs ConditionalKeys

I'm uncertain about what approach is better but it seems that `ConditionalKeys` proposed in #18967 will filter off more unapplicable keys from Chainable. I don't think there is a much difference for the consumer though.

**Comment by xumepadismal** [comment]
I can modify my PR to take into account all the goodies from #18967 and #18880 if this is acceptable and nobody minds :)

**Comment by remcohaszing** [comment]
I think this PR does everything #18967 does plus more (`prevSubject`), so I recommend to close that one and merge this instead.

> ### Regarding the void
> In #18967 return type of `fn` is just `void` which is not accurate. We should allow `void` but not require it. In my PR it is `ReturnType<ChainableMethods[T]> | void` which is better correspond the docs.

I used `void` because it turned out a return type of `void | Whatever` is equal to a return type of `void` for callback functions. Technically the unwrapped value of the chainable is also a valid return type.

**Comment by xumepadismal** [comment]
@remcohaszing 
> I used `void` because it turned out a return type of `void | Whatever` is equal to a return type of `void` for callback functions.

Hmm I played around this and it seems that TS checks the return type  correctly, no? You can return either correct type or nothing. Error if you try to return something inconsistent to interface.

image

**Comment by xumepadismal** [comment]
Now this PR should also cover everything in #18880. Hopefully I didn't miss anything, review will be welcome :)

**Review by SimeonC — DISMISSED** [review]
Looks like this covers all the cases from my PR #18880 , thanks 👍 

Could you add that this closes #18879 as well?

File: cli/types/cypress.d.ts [hunk 1]
```diff
@@ -420,9 +445,17 @@ declare namespace Cypress {
...
-      add<T extends keyof Chainable>(name: T, options: CommandOptions, fn: Chainable[T]): void
-      overwrite<T extends keyof Chainable>(name: T, fn: Chainable[T]): void
+      add<T extends keyof Chainable>(name: T, fn: CommandFn<T>): void
+      add<T extends keyof Chainable>(name: T, options: CommandOptions & {prevSubject: false}, fn: CommandFn<T>): void
+      add<T extends keyof Chainable>(name: T, options: CommandOptions & {prevSubject: true}, fn: CommandFnWithSubject<T, unknown>): void
+      add<T extends keyof Chainable, S extends PrevSubject>(
+          name: T, options: CommandOptions & { prevSubject: S | Array<Exclude<S, 'optional'>> }, fn: CommandFnWithSubject<T, PrevSubjectMap[S]>,
+      ): void
+      add<T extends keyof Chainable, S extends PrevSubject>(
+          name: T, options: CommandOptions & { prevSubject: 'optional' | ['optional', ...S[]] }, fn: CommandFnWithSubject<T, void | PrevSubjectMap[S]>,
```
Comment (hgabor995) [inline cli/types/cypress.d.ts:455]: Will this section work fine even if 'optional' isn't in the first place?

Comment (xumepadismal) [reply cli/types/cypress.d.ts]: Just pushed a commit to support this case. I'm little bit surprised the code looks even cleaner with this fix :)

**Review by dclowd9901 — APPROVED** [review]

**Review by mjhenkes — APPROVED** [review]

**Comment by mrkjdy** [comment]
I'm not sure if this is the right place to post this, but I got a chance to checkout the changes in the PR, and when overwriting commands like `'select'` it's not obvious that you need to add type parameters to overwrite like this in order for the correct types to be picked up:
```TypeScript
Cypress.Commands.overwrite<'select', 'element'>('select', (originalFn, prevSubject, valueOrTextOrIndex, options) => {
  // ...
});
```

After this PR is merged can [the docs](https://docs.cypress.io/guides/tooling/typescript-support#Types-for-custom-commands) be updated to mention this?

**Review by sagarrajput — APPROVED** [review]

**Comment by chrisbreiding** [comment]
I agree with @mrkjdy that the docs should be updated to reflect these changes. We prefer to do docs updates in tandem with the code updates and then release them together.

@xumepadismal Would you mind opening a PR in the [cypress-documentation](https://github.com/cypress-io/cypress-documentation) updating the [typescript support doc](https://github.com/cypress-io/cypress-documentation/blob/master/content/guides/tooling/typescript-support.md#types-for-custom-commands)?

**Comment by xumepadismal** [comment]
> I agree with @mrkjdy that the docs should be updated to reflect these changes. We prefer to do docs updates in tandem with the code updates and then release them together.
> 
> @xumepadismal Would you mind opening a PR in the [cypress-documentation](https://github.com/cypress-io/cypress-documentation) updating the [typescript support doc](https://github.com/cypress-io/cypress-documentation/blob/master/content/guides/tooling/typescript-support.md#types-for-custom-commands)?

Sorry very busy these days but I'll try to allocate some time till the end of the week to create a PR in the docs

**Comment by nitzanashi** [comment]
may I ask why the Subject is set to be `unknown` now?

**Comment by NicholasBoll** [comment]
This PR introduces issues with `Cypress.Commands.overwrite` on commands with overloads. There is no way for Typescript to know which overload should be used, so the last overload is resolved. Overall, the PS is nice for `Cypress.Commands.add` to remove duplicate typing, but it fails for `Cypress.Commands.overwrite`. If there is no overloads for the command, it works fine and is nice to have `originalFn` properly typed, but there is no way to choose the overload. I've had to be clever with a `visit` override to do something on a `beforeload` event:

Example:

```ts
Cypress.Commands.overwrite('visit', (originalFn, urlOrOptions, inputOptions = {}) => {
  let options: typeof urlOrOptions; // weird. Typescript doesn't think `urlOrOptions` could be a `string`

  // usually these types of guards are type guards, but now this is only a runtime guard  
  if (typeof urlOrOptions === 'object') {
    options = urlOrOptions;
  } else {
    // wait, `urlOrOptions` is of type `never` here...
    options = {url: urlOrOptions, ...inputOptions};
  }

  return originalFn({
    ...options,
    onBeforeLoad(win) {
      options.onBeforeLoad?.(win);
      supports(); // prime the ally.js supports cache so it doesn't mess with the cypress-plugin-tab
    },
  });
});
```

In this example, `urlOrOptions` is `string | Partial<Cypress.VisitOptions> & { url: string }` because of the overload, but since the last overload is taken, Typescript now thinks the type is never a `string`. I can no longer override parameters to type them correctly for internal use unless I cast `(Cypress.Commands.overwrite as any)` first.

**Comment by NicholasBoll** [comment]
Also slightly nuanced is the return type of the original function:

```ts
Cypress.Commands.overwrite('visit', (originalFn, options) => {
  const foo = originalFn(options);

  return foo
});
```

In this example, `foo` is a Bluebird promise, but the types say it is `Cypress.Chainable<Cypress.AUTWindow>`. I ran into this issue with some overrides:

```ts
return Cypress.Promise.try(() => {
  return originalFn(subject, options)
})
.then(value => {
  // value should be `JQuery<HTMLElement>` since the original fn returns a jQuery element
  // but `value` is typed as `Cypress.Chainable` since Bluebird cannot unwrap the subject because the way this is typed
  // says `originalFn` returns `Cypress.Chainable<Subject>` instead of `Subject` or `Bluebird.Promise<Subject>`
})
```

The problem is Cypress commands normalized around a `Cypress.Chainable` return type, but the original functions do not have any such normalization.

**Comment by elio-maggini** [comment]
Is there a time line on when this might be merged? Currently all custom commands we've created are non-functional.
